### PR TITLE
use secp256k1 ecdsa prehash with keccak256

### PR DESCRIPTION
**Pull request by semenov-vladyslav**
# Description of change

Please write a summary of your changes and why you made them.

## Links to any relevant issues

Be sure to reference any related issues by adding `fixes issue #`.

## Type of change

Choose a type of change, and delete any options that are not relevant.

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [ ] Enhancement (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Fix

## How the change has been tested

Describe the tests that you ran to verify your changes.

Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/signatures/secp256k1_ecdsa.rs [hunk 1]
```diff
@@ -38,20 +44,47 @@ impl SecretKey {
...
+    #[cfg(feature = "keccak")]
+    pub fn sign_keccak256(&self, msg: &[u8]) -> RecoverableSignature {
+        self.try_sign_keccak256(msg).expect("Secp256k1 ECDSA sign failed")
+    }
+
+    /// Generate Standard Secp256k1 ECDSA signature of SHA256 hash value of a message.
+    /// Signature generation can fail with a very low probability.
+    pub fn try_sign_sha256(&self, msg: &[u8]) -> crate::Result<RecoverableSignature> {
         self.0
             .sign_recoverable(msg)
```
Comment (thibault-martinez) [inline src/signatures/secp256k1_ecdsa.rs:74]: Just checking, this internally does the hash?

**Review by thibault-martinez — APPROVED** [review]

File: src/signatures/secp256k1_ecdsa.rs (hunk 1 above)
Comment (semenov-vladyslav) [reply src/signatures/secp256k1_ecdsa.rs:74]: Yes, hash function is resolved for the `Secp256k1` type [here](https://github.com/RustCrypto/elliptic-curves/blob/master/k256/src/ecdsa.rs#L182). To override it it's possible to use `sign_digest_recoverable` or [`sign_prehash_recoverable`](https://github.com/RustCrypto/signatures/blob/master/ecdsa/src/recovery.rs#L180).

**Review by thibault-martinez — APPROVED** [review]

File: src/signatures/secp256k1_ecdsa.rs [hunk 2]
```diff
@@ -38,20 +44,47 @@ impl SecretKey {
...
+    #[cfg(feature = "keccak")]
+    pub fn try_sign_keccak256(&self, msg: &[u8]) -> crate::Result<RecoverableSignature> {
+        let mut prehash = [0_u8; PREHASH_LENGTH];
+        keccak256(msg, &mut prehash);
+        self.try_sign_prehash(&prehash)
+    }
+
+    /// Secp256k1 ECDSA signature of Keccak256 hash value of a message as used in Ethereum.
+    #[cfg(feature = "keccak")]
+    pub fn sign_keccak256(&self, msg: &[u8]) -> RecoverableSignature {
```
Comment (Thoralf-M) [inline src/signatures/secp256k1_ecdsa.rs]: Why even have this function?

Comment (thibault-martinez) [reply src/signatures/secp256k1_ecdsa.rs]: I kinda agree, I'm not a big fan of functions that hide a panic for not much reason, maybe we should juste leave the user unwrap themselves. Or at least add documentation about the underlying panic.

Comment (semenov-vladyslav) [reply src/signatures/secp256k1_ecdsa.rs]: The probability of the event that signature generation fails is _very_ low. But it's not zero. Not clear how users should handle the error.

Comment (semenov-vladyslav) [reply src/signatures/secp256k1_ecdsa.rs]: Removed the functions.

**Review by Thoralf-M — APPROVED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**